### PR TITLE
LPE alt init now allows baro only init without GPS w/o changing flag.

### PR DESCRIPTION
**Pull request by jgoppert**
Still need to flight test.

### CODE REVIEW

**Comment by Alirezam96** [comment]
thank for your update.
i do that changes its going to altitude mode successfully without reject massage but if change mode to altitude vehicle is starting to reduce alt.

**Comment by jgoppert** [comment]
You need to center your throttle stick after switching to altitude hold. Your plot shows your stick was around 40%. It needs to be at 50%.

**Comment by kd0aij** [comment]
@Alirezam96 If you set parameter MPC_THR_HOVER to 0.4, you should find that hover is at center stick in manual mode. Then when you switch to ALTCTL while hovering, your vertical speed should remain at zero. Unless there's a bug in the initialization of the altitude setpoint.

**Comment by jgoppert** [comment]
@kd0aij I flew my iris last night with latest commit to master and did the same thing, set the MPC_THR_HOVER to 0.75 since I had it loaded up, flew well in altitude hold mode.

**Comment by kd0aij** [comment]
@jgoppert Thanks, that's a pretty definitive test. Lately it seem to me that there's no deadband at all on the throttle stick in ALTCTL, even though I have MPC_ALTCTL_DZ set to 0.1 and MPC_ALTCTL_DY set to zero. Is there some other parameter that might affect this?

**Comment by jgoppert** [comment]
@kd0aij I agree. The deadband on my tarranis is only a few click of the throttle stick. It seems like the deadband isn't actually 0.1.

**Comment by kd0aij** [comment]
OK, I'll see if I can find the problem

**Comment by Alirezam96** [comment]
MPC_THR_HOVER is 50% and my throttle stick is around 50% but still start to reduce altitude.
SENS.BaroAlt data is almost constant in altitude estimate plot from log file.
i think that barometer isn't working well or baro data doesnt ok for controller.

**Comment by kd0aij** [comment]
The throttle deadband is implemented correctly for ALTCTL mode, but when the manual command is for zero velocity, MPC switches from z-velocity control to to "alt_control". If pos_sp(2) is drifting, that would explain the nonzero vertical speed at center stick. I suppose it was assumed that the altitude estimate is more accurate than the velocity estimate, and it seems that this should be true for filtered baro altitude, though I doubt it for GPS altitude.

**Comment by kd0aij** [comment]
@Alirezam96 Your log shows your RC throttle *never* reaching 50%

**Comment by Alirezam96** [comment]
Excuse me but do you know which parameter is altitude controller use??

**Comment by kd0aij** [comment]
The throttle channel controls vertical speed in altctl mode. In your log that is channel RC.C2: http://logs.uaventure.com/view/87pGw4qNqVuq9qZMQmQruF#RC_Raw_PLOT

**Comment by Alirezam96** [comment]
Understand and calib RC several times but still mid throttle in transmitter isn't 50% throttle in board what should i do?

**Comment by kd0aij** [comment]
Did you see this? https://github.com/PX4/Firmware/pull/5398#issuecomment-242976525

If your copter doesn't hover at mid-stick, adjust MPC_THR_HOVER till it does. To test, put the copter in a stable hover in manual mode, then switch to ALTCTL. If you have it right, the copter will not change altitude.

**Comment by Alirezam96** [comment]
Got it.

**Comment by kd0aij** [comment]
@jgoppert I just flew latest master on both fmu-v4 and fmu-v2 and verified that the deadband in ALTCTL mode is correct using LPE with GPS disabled. So if there is a problem, it must involve use of GPS data.

**Comment by Alirezam96** [comment]
I did all thing that you said and my hover throttle is around 50% but i think sth that dont know is reduce altitude set point so altitude start to decrease.

**Comment by kd0aij** [comment]
could you rephrase that? What about the altitude setpoint?

**Comment by Alirezam96** [comment]
I think the reason of reducing altitude is decreasing alt setpoint.

**Comment by kd0aij** [comment]
That won't happen in ALTCTL if your throttle stick is within the deadband, default 40% to 60%. Is your RC throttle at 50% when the altitude is decreasing?

**Comment by Alirezam96** [comment]
yes as you can see in http://logs.uaventure.com/view/mi2W93FTVsqHAKCpiHitw3

**Comment by kd0aij** [comment]
I don't see GPS or lidar data in that log, and the baro alt looks constant. What sensor is providing the altitude measurement?

**Comment by kd0aij** [comment]
correction, I can see the baro is working using Flightplot. What kind of accuracy are you expecting from baro altitude?

**Comment by kd0aij** [comment]
@jgoppert In that last log, LPSP.Z is tracking LPOS.Z; it looks like even though nav_state is 1, there's no altitude control loop running

**Comment by kd0aij** [comment]
@Alirezam96 I noticed you have MPC_ALTCTL_DY set to 0.1 instead of 0.0. Did you ever try zero, as I suggested?

**Comment by Alirezam96** [comment]
I know its working but I think there is no altitude control so as you see it start to decrease alt.
I got very good result from pixhawk baro with ardupilot v3.3.3 firmware.
that firmware is not working well in pixracer and px4 firmware have these problems.

**Comment by kd0aij** [comment]
best of luck

**Comment by kd0aij** [comment]
@Alirezam96 you might try this PR: https://github.com/PX4/Firmware/pull/5409 to see if it fixes your problem
